### PR TITLE
gRPC stabilization

### PR DESCRIPTION
**Pull request by ggezer**
**Completed:**
- Clients don't send defaults to FileSystem methods
- Selectively disabling DFSIO test based on hadoop client version
- Removed deprecated FSM Rest handler
- Deprecated key/value services
- Guava bumped to latest
- Merge transport and transport-options
- Hardcoded gRPC java plugin
- TODO cleanup

### CODE REVIEW

**Comment by aaudiber** [comment]
Jenkins, test this please

**Comment by aaudiber** [comment]
Jenkins, test this please

**Comment by calvinjia** [comment]
@ggezer very cool, thanks.

File: tests/src/test/java/alluxio/client/hadoop/DFSIOIntegrationTest.java [hunk 1]
```diff
@@ -87,10 +91,29 @@
...
+      return new Statement() {
+        @Override
+        public void evaluate() throws Throwable {
+          String hadoopVersion = System.getProperty("alluxio.hadoop.version");
+          if (hadoopVersion != null && (hadoopVersion.startsWith("2.4")
+              || hadoopVersion.startsWith("2.5") || hadoopVersion.startsWith("2.6"))) {
+            throw new AssumptionViolatedException("Hadoop version not supported. Skipping test!");
+          } else {
+            base.evaluate();
+          }
```
Comment (calvinjia) [inline tests/src/test/java/alluxio/client/hadoop/DFSIOIntegrationTest.java]: Can you use `AssumeFalse` here?

File: core/common/src/test/java/alluxio/AbstractClientTest.java [hunk 2]
```diff
@@ -79,7 +79,7 @@ public void unsupportedVersion() throws Exception {
     client.close();
   }
 
-  // TODO(ggezer) Fix after gRPC initialization
+  // TODO(ggezer) Fix
```
Comment (calvinjia) [inline core/common/src/test/java/alluxio/AbstractClientTest.java]: Could you @Ignore the tests instead of commenting them out?

File: core/common/src/main/java/alluxio/PropertyKey.java [hunk 3]
```diff
@@ -3859,6 +3868,8 @@ private static String javadocLink(String fullyQualifiedClassname) {
     public static final String SECURITY_LOGIN_IMPERSONATION_USERNAME =
         "alluxio.security.login.impersonation.username";
     public static final String SECURITY_LOGIN_USERNAME = "alluxio.security.login.username";
+    public static final String AUTHENTICATION_STALE_CHANNEL_PURGE_INTERVAL =
+        "alluxio.security.purge.interval";
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/PropertyKey.java]: Could you make the configuration name more explicit?

File: tests/src/test/java/alluxio/server/ft/journal/JournalIntegrationTest.java [hunk 4]
```diff
@@ -130,7 +131,11 @@ public void run() {
         } catch (UnavailableException e) {
           // Ignore - this happen if the client is interrupted while trying to connect.
         } catch (IOException | AlluxioException e) {
-          mException.set(e);
+          if (e.getCause() instanceof CanceledException) {
```
Comment (calvinjia) [inline tests/src/test/java/alluxio/server/ft/journal/JournalIntegrationTest.java]: I feel like we should be able to determine if we can ignore the exception by just inspecting the type of the exception. Unrelated to this PR if we can't.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 5]
```diff
@@ -3381,8 +3381,10 @@ private void setAttributeSingleFile(RpcContext rpcContext, LockedInodePath inode
     }
     UpdateInodeEntry.Builder entry = UpdateInodeEntry.newBuilder().setId(inode.getId());
     if (protoOptions.hasReplicationMax() || protoOptions.hasReplicationMin()) {
-      Integer replicationMax = protoOptions.getReplicationMax();
-      Integer replicationMin = protoOptions.getReplicationMin();
+      Integer replicationMax =
+          protoOptions.hasReplicationMax() ? protoOptions.getReplicationMax() : null;
+      Integer replicationMin =
+          protoOptions.hasReplicationMin() ? protoOptions.getReplicationMin() : null;
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Could this logic be written in a more concise way?

**Comment by calvinjia** [comment]
@ggezer Thanks for the clean up, I left a few comments

File: tests/src/test/java/alluxio/client/cli/fs/AbstractShellIntegrationTest.java [hunk 6]
```diff
@@ -34,7 +34,8 @@
       new LocalAlluxioClusterResource.Builder()
           .setProperty(PropertyKey.WORKER_MEMORY_SIZE, SIZE_BYTES)
           .setProperty(PropertyKey.USER_BLOCK_SIZE_BYTES_DEFAULT, SIZE_BYTES)
-          .setProperty(PropertyKey.MASTER_TTL_CHECKER_INTERVAL_MS, Integer.MAX_VALUE).build();
+          .setProperty(PropertyKey.MASTER_TTL_CHECKER_INTERVAL_MS, Integer.MAX_VALUE)
+          .setProperty(PropertyKey.USER_FILE_WRITE_TYPE_DEFAULT, "CACHE_THROUGH").build();
```
Comment (gpang) [inline tests/src/test/java/alluxio/client/cli/fs/AbstractShellIntegrationTest.java]: why is this added?

File: core/client/fs/src/main/java/alluxio/client/file/options/InStreamOptions.java [hunk 7]
```diff
@@ -55,11 +57,24 @@ public InStreamOptions(URIStatus status) {
    * @param options OpenFile options
    */
   public InStreamOptions(URIStatus status, OpenFilePOptions options) {
+    // Create OpenOptions builder with default options.
```
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/options/InStreamOptions.java]: Is this missing the `USER_FILE_WRITE_LOCATION_POLICY` default?

File: core/client/fs/src/main/java/alluxio/client/file/options/OutStreamOptions.java [hunk 8]
```diff
@@ -69,19 +70,35 @@ public static OutStreamOptions defaults() {
...
+    if (options.hasReplicationMin()) {
+      mReplicationMin = options.getReplicationMin();
+    }
+    if (options.hasReplicationMax()) {
+      mReplicationMax = options.getReplicationMax();
+    }
+    if (options.hasWriteTier()) {
+      mWriteTier = options.getWriteTier();
+    }
+    if (options.hasWriteType()) {
```
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/options/OutStreamOptions.java]: If the options are not set, do these member variables get set correctly?

File: core/client/fs/src/test/java/alluxio/client/file/FileSystemTestUtils.java [hunk 9]
```diff
@@ -71,7 +71,7 @@ public static void createByteFile(FileSystem fs, String fileName,
   public static void createByteFile(FileSystem fs, AlluxioURI fileURI,
       WritePType writeType, int len) {
     CreateFilePOptions options =
-        FileSystemClientOptions.getCreateFileOptions().toBuilder().setWriteType(writeType).build();
+        CreateFilePOptions.newBuilder().setRecursive(true).setWriteType(writeType).build();
```
Comment (gpang) [inline core/client/fs/src/test/java/alluxio/client/file/FileSystemTestUtils.java]: why is recursive set now?

File: core/client/fs/src/test/java/alluxio/client/file/FileSystemTestUtils.java [hunk 10]
```diff
@@ -107,9 +107,8 @@ public static void createByteFile(FileSystem fs, AlluxioURI fileURI, CreateFileP
    */
   public static void createByteFile(FileSystem fs, String fileName, WritePType writeType, int len,
       long blockCapacityByte) {
-    CreateFilePOptions options = FileSystemClientOptions.getCreateFileOptions().toBuilder()
-        .setWriteType(writeType).setBlockSizeBytes(blockCapacityByte)
-        .build();
+    CreateFilePOptions options = CreateFilePOptions.newBuilder().setWriteType(writeType)
+        .setBlockSizeBytes(blockCapacityByte).setRecursive(true).build();
```
Comment (gpang) [inline core/client/fs/src/test/java/alluxio/client/file/FileSystemTestUtils.java]: why is recursive set now?

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 11]
```diff
@@ -135,7 +134,9 @@ public FSDataOutputStream append(Path path, int bufferSize, Progressable progres
       if (mFileSystem.exists(uri)) {
         throw new IOException(ExceptionMessage.FILE_ALREADY_EXISTS.getMessage(uri));
       }
-      return new FSDataOutputStream(mFileSystem.createFile(uri), mStatistics);
+      return new FSDataOutputStream(
+          mFileSystem.createFile(uri, CreateFilePOptions.newBuilder().setRecursive(true).build()),
```
Comment (gpang) [inline core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: why is recursive set now?

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 12]
```diff
@@ -175,8 +176,8 @@ public FSDataOutputStream create(Path path, FsPermission permission, boolean ove
     }
 
     AlluxioURI uri = new AlluxioURI(HadoopUtils.getPathWithoutScheme(path));
-    CreateFilePOptions options = FileSystemClientOptions.getCreateFileOptions().toBuilder()
-        .setBlockSizeBytes(blockSize).setMode(permission.toShort()).build();
+    CreateFilePOptions options = CreateFilePOptions.newBuilder().setBlockSizeBytes(blockSize)
+        .setMode(permission.toShort()).setRecursive(true).build();
```
Comment (gpang) [inline core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: why is recursive set now?

File: core/common/src/main/java/alluxio/PropertyKey.java [hunk 13]
```diff
@@ -3062,6 +3063,14 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.ENFORCE)
           .setScope(Scope.CLIENT)
           .build();
+  public static final PropertyKey AUTHENTICATION_STALE_CHANNEL_PURGE_INTERVAL =
+      new Builder(Name.AUTHENTICATION_STALE_CHANNEL_PURGE_INTERVAL)
+          .setDefaultValue("60min")
+          .setDescription("Interval at which stale channels will be removed from "
+              + "authentication cache.")
```
Comment (gpang) [inline core/common/src/main/java/alluxio/PropertyKey.java]: Could you also briefly mention the implications of shorter and longer intervals? This description will be in the documentation, so how/when would you advise users to adjust this?

File: core/common/src/main/java/alluxio/security/authentication/AuthenticatedUserInjector.java [hunk 14]
```diff
@@ -64,11 +65,15 @@ public void onHalfClose() {
...
           } catch (UnauthenticatedException e) {
             call.close(Status.UNAUTHENTICATED, headers);
+            callClosed = true;
           }
         } else {
           call.close(Status.UNAUTHENTICATED, headers);
+          callClosed = true;
+        }
+        if (!callClosed) {
+          super.onHalfClose();
```
Comment (gpang) [inline core/common/src/main/java/alluxio/security/authentication/AuthenticatedUserInjector.java]: could add a comment as to why `super.onHalfClose()` has to be called, in these circumstances?

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 5 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: protobuf defaults these fields to 0 which has meaning for the replication logic. Invoked code expects only explicitly being set values to be non-null, so there isn't much I can do at this layer.

File: core/common/src/main/java/alluxio/PropertyKey.java (hunk 3 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/PropertyKey.java]: I thought about it and came up with the same name again :)
Any suggestions?

File: core/base/src/main/java/alluxio/file/options/CompleteUfsFileOptions.java [hunk 15]
```diff
@@ -98,7 +99,7 @@ public int hashCode() {
 
   @Override
   public String toString() {
-    return Objects.toStringHelper(this)
+    return MoreObjects.toStringHelper(this)
```
[the remark this answers is not in the document]
Comment (ggezer) [reply core/base/src/main/java/alluxio/file/options/CompleteUfsFileOptions.java]: Due to Guava bump.

File: core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java [hunk 16]
```diff
@@ -130,13 +130,8 @@ public FileOutStream createFile(AlluxioURI path, CreateFilePOptions options)
     try {
       masterClient.createFile(path, options);
       // Do not sync before this getStatus, since the UFS file is expected to not exist.
-      GetStatusPOptions gsOptions = FileSystemClientOptions.getGetStatusOptions();
-      gsOptions = gsOptions.toBuilder()
-          .setLoadMetadataType(LoadMetadataPType.NEVER)
-          .setCommonOptions(
-              FileSystemClientOptions.getCommonOptions().toBuilder()
-                  .setSyncIntervalMs(-1))
```
[the remark this answers is not in the document]
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java]: It was redundant as -1 is the master default.

File: core/common/src/main/java/alluxio/PropertyKey.java (hunk 3 above)
Comment (calvinjia) [reply core/common/src/main/java/alluxio/PropertyKey.java]: my concern is the current name is too general, it could be a purge interval for credentials for instance. Maybe `alluxio.security.stale.channel.purge.interval` or some version of that which does not have as many dots.

File: core/common/src/test/java/alluxio/ConfigurationTestUtils.java [hunk 17]
```diff
@@ -106,7 +106,7 @@ public static void resetConfiguration() {
     // default write type becomes MUST_CACHE, set this value to CACHE_THROUGH for tests.
     // default Alluxio storage is STORE, and under storage is SYNC_PERSIST for tests.
     // TODO(binfan): eliminate this setting after updating integration tests
-    conf.put(PropertyKey.USER_FILE_WRITE_TYPE_DEFAULT, "CACHE_THROUGH");
+    //conf.put(PropertyKey.USER_FILE_WRITE_TYPE_DEFAULT, "CACHE_THROUGH");
```
[the remark this answers is not in the document]
Comment (ggezer) [reply core/common/src/test/java/alluxio/ConfigurationTestUtils.java]: It was set to CACHE_THROUGH to circumvent "something" which is not required after stopping to send client defaults.
I was going to either set it to CACHE_THROUGH in some tests, OR,  set explicitly to the default (MUST_CACHE) in many test.  I complied with the TODO just above and removed the test specific override.

File: tests/src/test/java/alluxio/server/tieredstore/TierPromoteIntegrationTest.java [hunk 18]
```diff
@@ -98,13 +99,16 @@ public void promoteBlock() throws Exception {
     AlluxioURI path3 = new AlluxioURI(PathUtils.uniqPath());
 
     // Write three files, first file should be in ssd, the others should be in memory
-    FileOutStream os1 = mFileSystem.createFile(path1);
+    FileOutStream os1 =
+        mFileSystem.createFile(path1, CreateFilePOptions.newBuilder().setRecursive(true).build());
```
[the remark this answers is not in the document]
Comment (ggezer) [reply tests/src/test/java/alluxio/server/tieredstore/TierPromoteIntegrationTest.java]: This is due to a behavior change in Alluxio Java API.
Earlier createFile was recursive by default, I removed that default. Which left many test code that just create a file without creating the directory structure first.

File: tests/src/test/java/alluxio/client/cli/fs/AbstractShellIntegrationTest.java (hunk 6 above)
Comment (ggezer) [reply tests/src/test/java/alluxio/client/cli/fs/AbstractShellIntegrationTest.java]: I explained this on the other comment.

File: tests/src/test/java/alluxio/client/hadoop/DFSIOIntegrationTest.java (hunk 1 above)
Comment (calvinjia) [reply tests/src/test/java/alluxio/client/hadoop/DFSIOIntegrationTest.java]: Got it, thanks.

File: core/client/fs/src/main/java/alluxio/client/file/options/InStreamOptions.java (hunk 7 above)
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/file/options/InStreamOptions.java]: OpenPOptions may have not contain a value for it due to not requiring client defaults.

InStreamOptions and OutStreamOptions are not compatible with this behavior change. These two actually requires the defaults so they read the Configuration object to get the defaults and then merge these with the given options object in case some have been set explicitly.

File: core/client/fs/src/main/java/alluxio/client/file/options/OutStreamOptions.java (hunk 8 above)
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/file/options/OutStreamOptions.java]: Yes, the OutStreamOptions::_ctor reads from Configuration.

File: core/client/fs/src/test/java/alluxio/client/file/FileSystemTestUtils.java (hunk 9 above)
Comment (ggezer) [reply core/client/fs/src/test/java/alluxio/client/file/FileSystemTestUtils.java]: Alluxio Java API change. Explained in other comment.

File: core/client/fs/src/test/java/alluxio/client/file/FileSystemTestUtils.java (hunk 10 above)
Comment (ggezer) [reply core/client/fs/src/test/java/alluxio/client/file/FileSystemTestUtils.java]: Alluxio Java API change. Explained in other comment.

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 11 above)
Comment (ggezer) [reply core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: Alluxio Java API change. Explained in other comment.

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 12 above)
Comment (ggezer) [reply core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: Alluxio Java API change. Explained in other comment.

File: tests/src/test/java/alluxio/server/ft/journal/JournalIntegrationTest.java (hunk 4 above)
Comment (ggezer) [reply tests/src/test/java/alluxio/server/ft/journal/JournalIntegrationTest.java]: Agreed.

Filtered CANCELLED status in the AbstractClient to throw UnavailableException.

File: core/common/src/main/java/alluxio/PropertyKey.java (hunk 13 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/PropertyKey.java]: I'd like to address this as part of merge PR if it's OK.
